msgpack/src/objects.cpp

```diff
@@ -0,0 +1,160 @@
+    return objects;
+}
+
+UserType::ControlFuncs* get_cfs_sbuffers()
+{
+    static UserType::ControlFuncs cfs;
+    cfs.ostream_func = [](std::ostream& os, const UserType& a) -> std::ostream& {
+        os << "msgpackSBuffer<" << a.data() << ">";
+        return os;
+    };
+
+    cfs.deleter = [](void* data) {
+        std::list<std::unique_ptr<msgpack::sbuffer>>& sbuffers = get_sbuffers();
+        for (auto it = sbuffers.begin(), end = sbuffers.end(); it != end; ++ it)
+        {
+            if (&(*it) == data)
+            {
+                sbuffers.erase(it);
+                break;
+            }
+        }
+    };
+
+    return &cfs;
+}
+
+UserType::ControlFuncs* get_cfs_obj_handlers()
+{
+    static UserType::ControlFuncs cfs;
+    cfs.ostream_func = [](std::ostream& os, const UserType& a) -> std::ostream& {
+        os << "msgpackObjectHandler<" << a.data() << ">";
+        return os;
+    };
+
+    cfs.deleter = [](void* data) {
+        std::list<std::unique_ptr<msgpack::object_handle>>& object_handlers = get_obj_handlers();
+        for (auto it = object_handlers.begin(), end = object_handlers.end(); it != end; ++ it)
+        {
+            if (&(*it) == data)
+            {
+                object_handlers.erase(it);
+                break;
+            }
+        }
+    };
+
+    return &cfs;
+}
+
+UserType::ControlFuncs* get_cfs_objects()
+{
+    static UserType::ControlFuncs cfs;
+    cfs.ostream_func = [](std::ostream& os, const UserType& a) -> std::ostream& {
+        os << "msgpackObject<" << a.data() << ">";
+        return os;
+    };
+
+    cfs.deleter = [](void* data) {
+        std::list<std::unique_ptr<msgpack::object>>& objects = get_objects();
+        for (auto it = objects.begin(), end = objects.end(); it != end; ++ it)
+        {
+            if (&(*it) == data)
+            {
+                objects.erase(it);
+                break;
+            }
+        }
+    };
+
+    return &cfs;
+}
+
+CObject get_cobject(Value& ark_object, ValueType type)
+{
+    CObject object;
+
+    switch(type)
+    {
+        case ValueType::True:
+            object = true;
+            break;
+        case ValueType::False:
+            object = false;
+            break;
+        case ValueType::Number:
+            object = ark_object.number();
+            break;
+        case ValueType::String:
+            object = ark_object.stringRef().toString();
+            break;
+        case ValueType::List:
+            object = ark_object.list();
+            break;
+    }
+
+    return object;
+}
+
+/*              Objects constructors                   */
+
+// msgpack::sbuffer
+Value msgpack_sbuffer(std::vector<Value>& args, VM* vm)
+{
+    if (args.size() != 0)
+        throw std::runtime_error("ArgError: No arguments necessary for this function");
+
+    get_sbuffers().emplace_back(std::make_unique<msgpack::sbuffer>());
+    Value sbuffer = Value(UserType(get_sbuffers().back().get()));
```

Comment:
Something you can do everywhere, since emplace_back returns a reference to the constructed object:

```cpp
Value buffer = Value(
    UserType(
        get_sbuffers().emplace_back(
            std::make_unique<msgpack::sbuffer>()
        ).get()));
```

Reply:
ok